File: apps/ios/Sources/Features/Channel/PricingSettings.swift
import Foundation

// MARK: - Domain Models

struct PricingSettings {
    let channelId: Int
    let pricingEnabled: Bool
    let defaultPrice: Int?
    let defaultPrices: [String: Int?]?
    let difficultyPrices: [String: Int?]?
    let difficultyPricesByCurrency: [String: [String: Int?]?]?
    let currencyUnit: String
    let currencyConfigs: [CurrencyConfig]
}

struct CurrencyConfig: Identifiable {
    var id: String { key }
    let key: String
    let unit: String
    let amount: Int?
}

// MARK: - DTOs

struct PricingSettingsDTO: Decodable {
    let channelId: Int?
    let pricingEnabled: Bool?
    let defaultPrice: Int?
    let defaultPrices: [String: Int?]?
    let difficultyPrices: [String: Int?]?
    let difficultyPricesByCurrency: [String: [String: Int?]?]?
    let currencyUnit: String?
    let currencyConfigs: [CurrencyConfigDTO]?

    func toDomain() -> PricingSettings {
        PricingSettings(
            channelId: channelId ?? 0,
            pricingEnabled: pricingEnabled ?? false,
            defaultPrice: defaultPrice,
            defaultPrices: defaultPrices,
            difficultyPrices: difficultyPrices,
            difficultyPricesByCurrency: difficultyPricesByCurrency,
            currencyUnit: currencyUnit ?? "",
            currencyConfigs: currencyConfigs?.map { $0.toDomain() } ?? []
        )
    }
}

struct CurrencyConfigDTO: Decodable {
    let key: String
    let unit: String
    let amount: Int?

    func toDomain() -> CurrencyConfig {
        CurrencyConfig(key: key, unit: unit, amount: amount)
    }
}

// MARK: - Request Payload

struct UpdatePricingSettingsPayload: Encodable {
    let pricingEnabled: Bool?
    let defaultPrice: Int?
    let defaultPrices: [String: Int?]?
    let difficultyPrices: [String: Int?]?
    let difficultyPricesByCurrency: [String: [String: Int?]?]?
    let currencyConfigs: [CurrencyConfigPayload]?

    init(
        pricingEnabled: Bool? = nil,
        defaultPrice: Int? = nil,
        defaultPrices: [String: Int?]? = nil,
        difficultyPrices: [String: Int?]? = nil,
        difficultyPricesByCurrency: [String: [String: Int?]?]? = nil,
        currencyConfigs: [CurrencyConfigPayload]? = nil
    ) {
        self.pricingEnabled = pricingEnabled
        self.defaultPrice = defaultPrice
        self.defaultPrices = defaultPrices
        self.difficultyPrices = difficultyPrices
        self.difficultyPricesByCurrency = difficultyPricesByCurrency
        self.currencyConfigs = currencyConfigs
    }
}

struct CurrencyConfigPayload: Encodable {
    let key: String
    let unit: String
    let amount: Int?
}

File: apps/android/app/src/main/java/chat/rogi/rogichat/channelport/core/model/song/PricingSettings.kt
package chat.rogi.rogichat.channelport.core.model.song

import kotlinx.serialization.Serializable

// MARK: - Domain Models

data class PricingSettings(
    val channelId: Int,
    val pricingEnabled: Boolean,
    val defaultPrice: Int?,
    val defaultPrices: Map<String, Int?>?,
    val difficultyPrices: Map<String, Int?>?,
    val difficultyPricesByCurrency: Map<String, Map<String, Int?>?>?,
    val currencyUnit: String,
    val currencyConfigs: List<CurrencyConfig>,
)

data class CurrencyConfig(
    val key: String,
    val unit: String,
    val amount: Int?,
)

// MARK: - DTOs

@Serializable
data class PricingSettingsDto(
    val channelId: Int = 0,
    val pricingEnabled: Boolean = false,
    val defaultPrice: Int? = null,
    val defaultPrices: Map<String, Int?>? = null,
    val difficultyPrices: Map<String, Int?>? = null,
    val difficultyPricesByCurrency: Map<String, Map<String, Int?>?>? = null,
    val currencyUnit: String = "",
    val currencyConfigs: List<CurrencyConfigDto> = emptyList(),
) {
    fun toDomain(): PricingSettings = PricingSettings(
        channelId = channelId,
        pricingEnabled = pricingEnabled,
        defaultPrice = defaultPrice,
        defaultPrices = defaultPrices,
        difficultyPrices = difficultyPrices,
        difficultyPricesByCurrency = difficultyPricesByCurrency,
        currencyUnit = currencyUnit,
        currencyConfigs = currencyConfigs.map { it.toDomain() },
    )
}

@Serializable
data class CurrencyConfigDto(
    val key: String,
    val unit: String,
    val amount: Int? = null,
) {
    fun toDomain(): CurrencyConfig = CurrencyConfig(
        key = key,
        unit = unit,
        amount = amount,
    )
}

@Serializable
data class UpdatePricingSettingsPayload(
    val pricingEnabled: Boolean? = null,
    val defaultPrice: Int? = null,
    val defaultPrices: Map<String, Int?>? = null,
    val difficultyPrices: Map<String, Int?>? = null,
    val difficultyPricesByCurrency: Map<String, Map<String, Int?>?>? = null,
    val currencyConfigs: List<CurrencyConfigDto>? = null,
)
